exe/kubernetes-deploy

```diff
@@ -293,23 +288,48 @@ class KubernetesDeploy
     end
   end
 
+  module DescendantsTracker
+    @@direct_descendants = {}
+
+    class << self
+      def direct_descendants(klass)
+        @@direct_descendants[klass] || []
+      end
+
+      def store_inherited(klass, descendant)
+        (@@direct_descendants[klass] ||= []) << descendant
+      end
+    end
+  end
+
   class KubernetesResource
-    extend ActiveSupport::DescendantsTracker
+    extend DescendantsTracker
 
     attr_reader :name, :namespace, :file
     attr_writer :type, :deploy_started
 
-    TIMEOUT = 5.minutes
+    TIMEOUT = 5 * 60
 
-    def self.handled_type
-      name.split('::').last
-    end
+    class << self
+      def direct_descendants
+        DescendantsTracker.direct_descendants(self)
+      end
 
-    def self.for_type(type, name, namespace, file)
-      if subclass = descendants.find { |subclass| subclass.handled_type.downcase == type }
-        subclass.new(name, namespace, file)
-      else
-        self.new(name, namespace, file).tap { |r| r.type = type }
+      def inherited(base)
+        DescendantsTracker.store_inherited(self, base)
+        super
+      end
+
+      def handled_type
+        name.split('::').last
+      end
+
+      def for_type(type, name, namespace, file)
+        if subclass = direct_descendants.find { |subclass| subclass.handled_type.downcase == type }
```

Comment:
I wouldn't say antipattern. But it's definitely better to strive for "strategy pattern" over iterating on all candidates to find what you are looking for.

Not that it makes a huge difference performance wise (since in this kind of cases, the candidates set is generally small), but it usually makes simpler, more direct code to read.

NB: I'm just answering this specific thread cause I saw some keywords that triggered me. But I haven't kept a close watch on that PR, so feel free to disregard my opinion.